sdk/identity/Azure.Identity/src/ManagedIdentityClient.cs

```diff
@@ -0,0 +1,419 @@
+            {
+                // aquire the init lock 
+                s_initLock.Wait(cancellationToken);
+
+                try
+                {
+                    // check again if the we already determined the msiType now that we hold the lock
+                    if (s_msiType == MsiType.Unknown)
+                    {
+                        string endpointEnvVar = Environment.GetEnvironmentVariable(MsiEndpointEnvironemntVariable);
+                        string secretEnvVar = Environment.GetEnvironmentVariable(MsiSecretEnvironemntVariable);
+
+                        // if the env var MSI_ENDPOINT is set
+                        if (!string.IsNullOrEmpty(endpointEnvVar))
+                        {
+                            s_endpoint = new Uri(endpointEnvVar);
+
+                            // if BOTH the env vars MSI_ENDPOINT and MSI_SECRET are set the MsiType is AppService
+                            if (!string.IsNullOrEmpty(secretEnvVar))
+                            {
+                                s_secret = secretEnvVar;
+                                s_msiType = MsiType.AppService;
+                            }
+                            // if ONLY the env var MSI_ENDPOINT is set the MsiType is CloudShell
+                            else
+                            {
+                                s_msiType = MsiType.AppService;
+                            }
+                        }
+                        // if MSI_ENDPOINT is NOT set AND the IMDS endpoint is available the MsiType is Imds
+                        else if (ImdsAvailable(cancellationToken))
+                        {
+                            s_endpoint = ImdsEndpoint;
+                            s_msiType = MsiType.Imds;
+                        }
+                        // if MSI_ENDPOINT is NOT set and IMDS enpoint is not available ManagedIdentity is not available
+                        else
+                        {
+                            s_msiType = MsiType.Unavailable;
+                        }
+                    }
+                }
+                // release the init lock
+                finally
+                {
+                    s_initLock.Release();
+                }
+            }
+
+            return s_msiType;
+        }
+
+        private bool ImdsAvailable(CancellationToken cancellationToken)
+        {            
+            // send a request without the Metadata header.  This will result in a failed request,
+            // but we're just interested in if we get a response before the timeout of 500ms
+            // if we don't get a response we assume the imds endpoint is not available
+            using (Request request = _pipeline.CreateRequest())
+            {
+                request.Method = HttpPipelineMethod.Get;
+
+                request.UriBuilder.Uri = ImdsEndpoint;
+
+                request.UriBuilder.AppendQuery("api-version", ImdsApiVersion);
```

Comment:
It's okay if we get a 400 due to no query string.

Reply:
I feel like specifying the API version is good.  Possibly future versions won't require the Metadata header.  I know that this is unlikely but adding the api-version is very low cost.

Reply:
Good point, I'll add it in the Python library as well.